src/providers/live/live-provider.ts

```diff
@@ -1,3 +1,222 @@
+    if (newTime === this._currentTime) {
+      return;
+    }
+    this._currentTime = newTime;
+    if (this._seekDifference !== null && this._currentTimeLive) {
+      // update _currentTimeLive after seek
+      this._currentTimeLive = this._currentTimeLive - this._seekDifference;
+    } else if (this._id3Timestamp) {
+      if (this._id3Timestamp === this._currentTimeLive) {
+        // prevent updating if calculated _currentTimeLive value the same as _id3Timestamp
+        this._id3Timestamp = null;
+        return;
+      }
+      // update _currentTimeLive from id3Tag time
+      this._currentTimeLive = this._id3Timestamp;
+    } else {
+      // update _currentTimeLive between id3Tags
+      this._currentTimeLive++;
+    }
+
+    this._id3Timestamp = null;
+    this._seekDifference = null;
+  };
+
+  private _handleSeeking = (): void => {
+    this._seekDifference = Math.ceil(this._currentTime - this._player.currentTime);
+  };
+
+  private _addBindings() {
+    this._eventManager.listen(this._player, this._player.Event.TIMED_METADATA, this._onTimedMetadataLoaded);
+    this._eventManager.listen(this._player, this._player.Event.SEEKING, this._handleSeeking);
+    this._eventManager.listen(this._player, this._player.Event.TIME_UPDATE, this._onTimeUpdate);
+  }
+
+  private _handleConnection = () => {
+    this._logger.debug('Connected to push server');
+  };
+
+  private _handleConnectionError = () => {
+    this._logger.error('Got an error during connection to push server');
+  };
+
+  private _initNotification(): void {
+    this._pushNotification.init();
+    this._constructPushNotificationListener();
+  }
+
+  private _prepareCuePoints<T extends ThumbPushNotificationData | SlideViewChangePushNotificationData>(cuePoint: T, cb: (cuePoint: T) => void) {
+    if (!this._player.currentTime) {
+      this._id3TagWaitTimeouts[cuePoint.id] = setTimeout(() => {
+        // wait till id3 tag handled
+        cb(cuePoint);
+      }, ID3_TAG_WAIT_TIMEOUT);
+      return;
+    }
+    delete this._id3TagWaitTimeouts[cuePoint.id];
+    cb(cuePoint);
+  }
+
+  private _fixCuePointEndTime<T extends ThumbPushNotificationData | SlideViewChangePushNotificationData>(cuePoints: T[]) {
+    return cuePoints
+      .sort((a, b) => {
+        return a.createdAt - b.createdAt;
+      })
+      .map((cue, index) => {
+        // fix endTime and replace VTTCue
+        if (cue.endTime === Number.MAX_SAFE_INTEGER && index !== cuePoints.length - 1) {
+          const fixedCue = {...cue, endTime: cuePoints[index + 1].startTime};
+          this._player.cuePointManager.addCuePoints([fixedCue]);
+        }
+        return cue;
+      });
+  }
+
+  private _prepareThumbCuePoints = (newThumb: ThumbPushNotificationData) => {
+    const startTime = this._player.currentTime - (this._currentTimeLive - newThumb.createdAt);
+    const ks = this._player.config.provider.ks;
+    const serviceUrl = this._player.config.provider.env.serviceUrl;
```

Comment:
see how and when vod provider take the ks and service url.
better to handle it in one place for both

Reply:
done